File: src/main/java/com/larryhsiao/clotho/source/ConstSource.java
package com.larryhsiao.clotho.source;

import com.larryhsiao.clotho.Source;

/**
 * Source holding a reference of given T.
 */
public class ConstSource<T> implements Source<T> {
    private final T ref;

    public ConstSource(T ref) {
        this.ref = ref;
    }

    @Override
    public T value() {
        return ref;
    }
}

File: src/main/java/com/larryhsiao/clotho/source/ConstSource.kt
package com.larryhsiao.clotho.source

import com.larryhsiao.clotho.Source

/**
 * Const source that only keeps the reference.
 * @todo #refactor-1 remove this class which is redundant
 */
class ConstSource<T>(private val value: T) : Source<T> {
    override fun value(): T {
        return value
    }
}